packages/rx_bloc_cli/mason_templates/bricks/rx_bloc_base/__brick__/lib/base/routers/router.dart

```diff
@@ -1,17 +1,232 @@
+import '../../feature_login/di/login_page_with_dependencies.dart';
+import '../../feature_notifications/di/notifications_page_with_dependencies.dart';
+import '../../feature_splash/di/splash_page_with_dependencies.dart';
+import '../common_blocs/coordinator_bloc.dart';
+import '../common_services/permissions_service.dart';
+import '../models/errors/error_model.dart';
+import '../models/item_model.dart';
+import 'go_router_refresh_stream.dart';
+import 'router_paths.dart';
+import 'router_permission_names.dart';
+
+part 'router.g.dart';
+
+class AppRouter {
+  AppRouter(this._context) {
+    refreshListener = GoRouterRefreshStream(
+        _context.read<CoordinatorBlocType>().states.isAuthenticated);
+  }
+
+  final BuildContext _context;
+  late GoRouterRefreshStream refreshListener;
+
+  GoRouter get router => _goRouter;
+
+  late final GoRouter _goRouter = GoRouter(
+    initialLocation: const SplashRoute().location,
+    routes: $appRoutes,
+    redirect: (context, state) async {
+      if (refreshListener.isLoggedIn && state.queryParams['from'] != null) {
+        return state.queryParams['from'];
+      }
+      if (refreshListener.isLoggedIn &&
+          state.subloc == const LoginRoute().location) {
+        return const CounterRoute().location;
+      }
+
+      final pathInfo =
+      router.routeInformationParser.matcher.findMatch(state.location);
+
+      final routeName = RouterPaths.pathToRouteName(pathInfo.fullpath);
+      final hasPermissions =
+      await context.read<PermissionsService>().hasPermission(routeName);
+
+      if (!refreshListener.isLoggedIn && !hasPermissions) {
+        return '${const LoginRoute().location}?from=${state.location}';
+      }
+
+      if (!hasPermissions) {
+        throw AccessDeniedErrorModel();
+      }
+
+      return null;
+    },
+    refreshListenable: GoRouterRefreshStream(
+      _context.read<CoordinatorBlocType>().states.isAuthenticated,
+    ),
+    errorPageBuilder: (context, state) => MaterialPage(
+      key: state.pageKey,
+      child: Scaffold(
+        appBar: AppBar(
+          title: const Text('Error page!'),
+        ),
+        body: const Center(
+          child: Text('Error message.'),
+        ),
+      ),
+    ),
+  );
+}
+
+abstract class RouteData {
+  String get permissionName;
+  String get routeLocation;
+}
+
+@TypedGoRoute<SplashRoute>(path: '/splash')
+@immutable
+class SplashRoute extends GoRouteData implements RouteData {
+  const SplashRoute();
+
+  @override
+  Page<Function> buildPage(BuildContext context, GoRouterState state) =>
+      MaterialPage(
+        key: state.pageKey,
+        child: const SplashPageWithDependencies(),
+      );
+
+  @override
+  String get permissionName => RouterPermissionNames.splashPage;
+
+  @override
+  String get routeLocation => location;
+}
+
+@TypedGoRoute<CounterRoute>(
+  path: '/',
+  routes: <TypedGoRoute<GoRouteData>>[
+    TypedGoRoute<NotificationsRoute>(path: 'notifications'),
```

Comment:
It's kind of confusing and error prone to define the paths in two places. Here and in `RouterPaths` as well

Reply:
reworked